projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-viewer.component.ts

```diff
@@ -0,0 +1,238 @@
+  @Input() autoplayRepeat = 1;
+
+  @Input() showControls = true;
+  @Input() showZoomControl = false; // zoom control allows the user to zoom into spectrogram
+  @Input() showAxisLabels = true;
+  @Input() axisFontSize = 10;
+  @Input() playOnlyOnSingleClick = false;
+
+  @Input() showPregeneratedSpectrogram = false;
+
+  @Input() spectrogramConfig: ISpectrogramConfig = {
+    sampleRate: 22050,
+    nperseg: 256,
+    noverlap: 256 - 160,
+    nbrOfRowsRemovedFromStart: 2,
+    maxNbrOfColsForNoiseEstimation: 6000,
+    noiseReductionParam: 2,
+    logRange: 3
+  };
+
+  @Input() mode: AudioViewerMode = 'default';
+
+  @Input() spectrogramWidth: number;
+  @Input() spectrogramHeight: number;
+  @Input() spectrogramMargin: { top: number, bottom: number, left: number, right: number };
+
+  @Input() audioInfoTpl: TemplateRef<any>;
+
+  @Input() zoomInfoText: string;
+
+  buffer: AudioBuffer;
+  audioPlayer: AudioPlayer;
+
+  loading = false;
+  hasError = false;
+
+  view: IAudioViewerArea;
+  defaultView: IAudioViewerArea;
+
+  @Output() audioLoading = new EventEmitter<boolean>();
+  @Output() drawEnd = new EventEmitter<IAudioViewerArea>();
+  @Output() spectrogramDblclick = new EventEmitter<number>();
+
+  private audioSub: Subscription;
+  private clicks = 0;
+
+  constructor(
+    private cdr: ChangeDetectorRef,
+    private audioService: AudioService,
+    private ngZone: NgZone
+  ) {
+    this.audioPlayer = new AudioPlayer(this.audioService, this.ngZone, this.cdr);
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.audio) {
+      this.clear();
+      this.setAudioLoading(true);
+
+      if (this.audio) {
+        this.audioSub = this.audioService.getAudioBuffer(this.audio.url, this.audio.duration).subscribe((buffer) => {
+          if (!this.areaIsValid(buffer, this.focusArea)) {
+            this.onError();
+            return;
+          }
+
+          this.buffer = buffer;
+          this.audioPlayer.setBuffer(buffer);
+          this.setDefaultView();
+
+          if (this.autoplay) {
+            this.audioPlayer.startAutoplay(this.autoplayRepeat);
+          }
+
+          this.cdr.markForCheck();
+        }, () => {
+          this.onError();
+        });
+      }
+    } else if (!this.hasError) {
+      if (changes.focusArea || changes.highlightFocusArea || changes.zoomTime || changes.timePaddingOnZoom || changes.zoomFrequency || changes.frequencyPaddingOnZoom) {
+        this.setDefaultView();
+      } else if (changes.mode) {
+        this.audioPlayer.stop();
+      }
+    }
+  }
+
+  ngOnDestroy() {
+    this.clear();
+  }
+
+  onSpectrogramDragStart() {
+    this.audioPlayer.stop();
+  }
+
+  onSpectrogramDragEnd(time: number) {
+    this.audioPlayer.startFrom(time);
+  }
+
+  onSpectrogramClick(time: number) {
+    if (!this.playOnlyOnSingleClick) {
+      this.audioPlayer.startFrom(time);
+    } else {
+      this.clicks++;
+      if (this.clicks === 1) {
+        setTimeout(() => {
```

Comment:
Why the timeout?

Reply:
It's a method for detecting that the click is a single click (and not a double-click). It waits a while and if the number of clicks have not increased, the click is determined to be a single click. I added some comments about this to the code.